workspace/src/path_planning/waypoints_path_planner/waypoints_path_planner/waypoints_path_planner.py

```diff
@@ -0,0 +1,221 @@
+        package_share_directory = get_package_share_directory("waypoints_path_planner")
+
+        self.declare_parameter("lookahead", 2.0)
+        self.lookahead = (
+            self.get_parameter("lookahead").get_parameter_value().double_value
+        )
+
+        # data that will be used by this class
+        self.state = VehicleState()
+
+        # subscribers
+        qos_profile = QoSProfile(depth=1)
+        qos_profile.history = QoSHistoryPolicy.KEEP_LAST
+        self.sub_state = self.create_subscription(
+            VehicleState, "~/input/vehicle_state", self.state_callback, qos_profile
+        )
+
+        # publishers
+        self.pub_err_state = self.create_publisher(
+            VehicleState, "~/output/error_state", 10
+        )
+        self.timer = self.create_timer(1 / self.freq, self.pub_callback)
+
+    # function to process data this class subscribes to
+    def state_callback(self, msg):
+        """The state Callback.
+
+        The callback to read the current state of the vehicle.
+
+        Args:
+            msg: The message received from the vehicle state topic
+
+        """
+        # self.get_logger().info("Received '%s'" % msg)
+        self.state = msg
+
+    def wpts_path_plan(self):
+        """The Waypoints path planner.
+
+        Computes the error state between the target state and the current state and the target state. First, get the target point which is closest ot the current state (plus the lookahead distance in the x direction). Then, get the error state by subtracting the current state from the target state (modular subtraction for the heading).
+
+        Returns:
+             The error state, and the target reference waypoint we are currently going to.
+
+        """
+        x_current = self.state.pose.position.x
+        y_current = self.state.pose.position.y
+        # theta_current = np.arctan2( 2*(x*w+z*y), x**2-w**2+y**2-z**2 )
+        theta_current = self.state.pose.orientation.z - 0.24845347641462115
+        while theta_current < -np.pi:
+            theta_current = theta_current + 2 * np.pi
+        while theta_current > np.pi:
+            theta_current = theta_current - 2 * np.pi
+
+        v_current = np.sqrt(
+            self.state.twist.linear.x**2 + self.state.twist.linear.y**2
+        )
+
+        # self.get_logger().info("SPEED: "+str(v_current))
+
+        dist = np.zeros((1, len(self.ref_traj[:, 1])))
+        for i in range(len(self.ref_traj[:, 1])):
+            dist[0][i] = dist[0][i] = (
+                x_current + np.cos(theta_current) * self.lookahead - self.ref_traj[i][0]
+            ) ** 2 + (
+                y_current + np.sin(theta_current) * self.lookahead - self.ref_traj[i][1]
+            ) ** 2
+            index = dist.argmin()
+
+        ref_state_current = list(self.ref_traj[index, :])
+        err_theta = 0
+        ref = ref_state_current[2]
+        act = theta_current
+
+        if (ref > 0 and act > 0) or (ref <= 0 and act <= 0):
```

Comment:
not a big problem here, will approve with it, but it's probs worth describing what these lines are doing

Reply:
Modular subtraction (want to be able to get the difference between angles in the right direction)

Reply:
will add a comment